src/Microsoft.FeatureManagement/FeatureFilters/Recurrence/RecurrenceEvaluator.cs

```diff
@@ -0,0 +1,695 @@
+        /// <param name="nextOccurrence">The next occurrence.</param>
+        /// <returns>True if the closest previous occurrence is within the recurrence range or the time is before the first occurrence, false otherwise.</returns>
+        /// </summary>
+        public static bool TryFindPrevAndNextOccurrences(DateTimeOffset time, TimeWindowFilterSettings settings, out DateTimeOffset prevOccurrence, out DateTimeOffset nextOccurrence)
+        {
+            prevOccurrence = DateTimeOffset.MinValue;
+
+            nextOccurrence = DateTimeOffset.MaxValue;
+
+            if (time < settings.Start.Value)
+            {
+                //
+                // The time is before the first occurrence.
+                nextOccurrence = settings.Start.Value;
+
+                return true;
+            }
+
+            if (TryFindPreviousOccurrence(time, settings, out prevOccurrence, out int numberOfOccurrences))
+            {
+                RecurrencePattern pattern = settings.Recurrence.Pattern;
+
+                switch (pattern.Type)
+                {
+                    case RecurrencePatternType.Daily:
+                        nextOccurrence = prevOccurrence.AddDays(pattern.Interval);
+
+                        break;
+
+                    case RecurrencePatternType.Weekly:
+                        nextOccurrence = GetWeeklyNextOccurrence(prevOccurrence, settings);
+
+                        break;
+
+                    default:
+                        return false;
+                }
+
+                RecurrenceRange range = settings.Recurrence.Range;
+
+                if (range.Type == RecurrenceRangeType.EndDate)
+                {
+                    if (nextOccurrence > range.EndDate)
+                    {
+                        nextOccurrence = DateTimeOffset.MaxValue;
+                    }
+                }
+
+                if (range.Type == RecurrenceRangeType.Numbered)
+                {
+                    if (numberOfOccurrences >= range.NumberOfOccurrences)
+                    {
+                        nextOccurrence = DateTimeOffset.MaxValue;
+                    }
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Perform validation of time window settings.
+        /// <param name="settings">The settings of time window filter.</param>
+        /// <param name="paramName">The name of the invalid setting, if any.</param>
+        /// <param name="reason">The reason that the setting is invalid.</param>
+        /// <returns>True if the provided settings are valid. False if the provided settings are invalid.</returns>
+        /// </summary>
+        public static bool TryValidateSettings(TimeWindowFilterSettings settings, out string paramName, out string reason)
```

Comment:
I follow the TargetingEvaluator.
https://github.com/microsoft/FeatureManagement-Dotnet/blob/preview/src/Microsoft.FeatureManagement/Targeting/TargetingEvaluator.cs#L239

Since the validation for recurrence settings is quite complex and there are a lot of methods for it. I agree to move them to another file.

Reply:
@rossgrambo Moved.